diff --git a/src/Service/CasUserManager.php b/src/Service/CasUserManager.php
index c48fc32..99abe93 100644
--- a/src/Service/CasUserManager.php
+++ b/src/Service/CasUserManager.php
@@ -163,11 +163,6 @@ class CasUserManager {
       }
     }

-    // Check if the retrieved user is blocked before moving forward.
-    if ($account->isBlocked()) {
-      throw new CasLoginException(sprintf('The username %s has not been activated or is blocked.', $account->getAccountName()), CasLoginExceptionType::AccountBlocked);
-    }
-
     // Dispatch an event that allows modules to prevent this user from logging
     // in and/or alter the user entity before we save it.
     $pre_login_event = new CasPreLoginEvent($account, $property_bag);
@@ -175,6 +170,14 @@ class CasUserManager {
     $this->eventDispatcher->dispatch($pre_login_event);
     $this->eventDispatcher->dispatch($pre_login_event, 'cas.pre_login');

+    // Reload the user account in case it has changed meanwhile.
+    $account = $this->externalAuth->load($property_bag->getUsername(), static::PROVIDER);
+
+    // Check if the retrieved user is blocked before moving forward.
+    if ($account->isBlocked()) {
+      throw new CasLoginException(sprintf('The username %s has not been activated or is blocked.', $account->getAccountName()), CasLoginExceptionType::AccountBlocked);
+    }
+
     // Save user entity since event listeners may have altered it.
     // @todo Don't take it for granted. Find if the account was really altered.
     // @todo Should this be swapped with the following if(...) block? Why
