diff --git a/src/Plugin/views/display/DataExport.php b/src/Plugin/views/display/DataExport.php
--- a/src/Plugin/views/display/DataExport.php
+++ b/src/Plugin/views/display/DataExport.php	(date 1749811236221)
@@ -942,6 +942,7 @@

       // Require at least one module granting access and none denying access.
       if (!empty($headers) && !in_array(-1, $headers)) {
+        \Drupal::service('eic_dashboards.export_log')->log($results);
         // Create a web server accessible URL for the private file.
         // Permissions for accessing this URL will be inherited from the View
         // display's configuration.
